control/freqplot.py

```diff
@@ -1039,14 +1040,173 @@ def gangof4_plot(P, C, omega=None, **kwargs):
+
+    Parameters
+    ----------
+    syslist : linsys
+        List of linear systems (single system is OK)
+    omega : array_like
+        List of frequencies in rad/sec to be used for frequency response
+    plot : bool
+        If True (default), plot magnitude and phase
+    omega_limits : array_like of two values
+        Limits of the to generate frequency vector.
+        If Hz=True the limits are in Hz otherwise in rad/s.
+    omega_num : int
+        Number of samples to plot.  Defaults to
+        config.defaults['freqplot.number_of_samples'].
+
+    Returns
+    -------
+    sigma : ndarray (or list of ndarray if len(syslist) > 1))
+        singular values
+    omega : ndarray (or list of ndarray if len(syslist) > 1))
+        frequency in rad/sec
+
+    Other Parameters
+    ----------------
+    grid : bool
+        If True, plot grid lines on gain and phase plots.  Default is set by
+        `config.defaults['bode.grid']`.
+
+    Examples
+    --------
+    >>> den = [75, 1]
+    >>> sys = ct.tf([[[87.8], [-86.4]], [[108.2], [-109.6]]], [[den, den], [den, den]])
+    >>> omega = np.logspace(-4, 1, 1000)
+    >>> sigma, omega = singular_values_plot(sys)
+
+    """
+    # Make a copy of the kwargs dictionary since we will modify it
+    kwargs = dict(kwargs)
+
+    # Check to see if legacy 'Plot' keyword was used
+    if 'Plot' in kwargs:
+        import warnings
+        warnings.warn("'Plot' keyword is deprecated in bode_plot; use 'plot'",
+                      FutureWarning)
+        # Map 'Plot' keyword to 'plot' keyword
+        plot = kwargs.pop('Plot')
+
+    # Get values for params (and pop from list to allow keyword use in plot)
+    dB = config._get_param('bode', 'dB', kwargs, _bode_defaults, pop=True)
+    Hz = config._get_param('bode', 'Hz', kwargs, _bode_defaults, pop=True)
+    grid = config._get_param('bode', 'grid', kwargs, _bode_defaults, pop=True)
+    plot = config._get_param('bode', 'grid', plot, True)
+
+    # If argument was a singleton, turn it into a tuple
+    if not hasattr(syslist, '__iter__'):
+        syslist = (syslist,)
+
+    # Decide whether to go above Nyquist frequency
+    omega_range_given = True if omega is not None else False
+
+    if omega is None:
+        omega_num = config._get_param(
+            'freqplot', 'number_of_samples', omega_num)
+        if omega_limits is None:
+            # Select a default range if none is provided
+            omega = _default_frequency_range(syslist, number_of_samples=omega_num)
+        else:
+            omega_range_given = True
+            omega_limits = np.asarray(omega_limits)
+            if len(omega_limits) != 2:
+                raise ValueError("len(omega_limits) must be 2")
+            if Hz:
+                omega_limits *= 2. * math.pi
+            omega = np.logspace(np.log10(omega_limits[0]),
+                                np.log10(omega_limits[1]), num=omega_num,
+                                endpoint=True)
```

Comment:
Very nice PR. I second @bnavigator's comment here. it would be great to have a common function to do what this code does, something analogous to the `_default_time_vector` private function in `timeresp.py` which would basically include this code. We could then also use it in the neglected `nichols` function as well (which should probably be put into `freqplot.py` IMO)

Reply:
Thanks @sawyerbfuller! Working on it :) BTW I am not sure whether the  _default_frequency_range as it is now is good for the svd plot also. On the distillation column example that I am testing the returned range seems to be too high-frequency!
Haven't looked inside this function yet. 

I'll do the refactoring anyways as as also  @bnavigator suggests (the rest of the logic is indeed the same for all the frequency-domain plots I guess)

Reply:
DONE! Now this logic is encapsulated in a private function:

omega_out, omega_range_given = _determine_frequency_range(syslist, omega_in, omega_limits, omega_num, Hz)

used by bode_plot and singular_values_plot. BTW, the _default_frequency_range function (which is invoked when the user
does not specify a range) it is perhaps not thought for a singular value plot and in some cases does not seem to produce a really "optimal" range. I think there is room for improvement in a future development...

Reply:
Sounds good. Perhaps a keyword could be added to `_default_frequency_range` in the future to request a range suitable for svd. 

Do you think it would make sense to change the name of the new function so that it is harder to confuse its purpose with that of `_default_frequency_range` ? One option would be something like `_default_omega_vector`, to parallel the equivalent function name in timeresp.py

Reply:
Name changed! Yes perhaps an option will be required if we come up with a better and specific logic for this plot.

Reply:
Yes a keyword may be the right way to do it if we find a better and specialized logic for the frequency range in 
the case of an svd plot. 

_default_omega_vector is also a bit misleading IMHO as the function computes the actually used range (the user-defined 
if specified, or the default auto range otherwise). The radix _default  is more appropriate for  the _default_frequency_range 
function I'd say.

I'd propose determine_omega_vector for this new function (or something else that gives the idea of a selection logic and not an autotuned range, I'm no mother-toungue)

Reply:
That name and reasoning sounds good to me.

Reply:
BTW the logic for the user-provided vector is not consistent I'd say:  if Hz=True, a user-provided omega_limits is  interpreted as Hz and multiplied times 2pi, but a user-provided omega is not modified.

Reply:
I wrote some of that code and it’s very possible that I didn’t think about that. So, probably a bug.

Reply:
Just a small detail :). 
The returned vector is always in rad/sec instead, irrespectively of the Hz parameter. This is a feature I guess, not a bug right?

Reply:
I think the consensus plan (not currently complete) is to have something like the following, arranged in terms of a pure math to engineering heirarchy: 

Sys(s/z) - done 

Sys.frequency_response(omega) -> (currrently returns mag and phase and should return sys(j*omega)  ) 

Sys.bode(omega) -> returns mag and phase, and can give omega in hz instead of rad/sec. plot optional. 

From that perspective, my take is that if Hz is true, then omega should be returned as hz too?

Reply:
Ah, I thought it was intended! 
As a user, I would also perhaps expect the output frequency to be in Hz (or at least in cycles/[UNIT_OF_TIME]) if I specify the Hz=True option, but this is not the case in the current implementation of bode (and in the one of the singular_value_plot in this PR).  Yet, I would perhaps expect the dB option to be for the plot only (don't know why, just my feeling!) 

Shall we deal with this issue in this PR, or keep it for another one?

Reply:
I pushed a new version where the passed omega is interpreted in Hz if Hz=True.
Note that I had to modify certain tests in config_test.py which were assuming instead omega to be still in rad/sec 
(and thus, Hz assumed to be only for plotting).  
Not sure whether there is really consensus on this point!

Reply:
> Not sure whether there is really consensus on this point!
That's a really good question!

The docstring still says 

```
    omega : array_like
        List of frequencies in rad/sec to be used for frequency response

    Returns
    -------
   ...
    omega : ndarray (or list of ndarray if len(syslist) > 1))
        frequency in rad/sec
```

If we deviate from that, we also have to adjust the docstring.

`omega` as opposed to `frequency`, `freq` or `f` also implies natural frequencies in rad/sec. All other frequency domain functions with input parameter `omega` expect rad/sec.

I vote to keep it strictly rad/sec.

Reply:
I would also prefer fixed units for input/outputs (for frequency, magnitude, and also for the phase in the  bode plot). 
Units options 'dB', 'Hz', and 'deg' then would only be used for the visualization. 
This interpretation is easier to document and less error-prone.

Reply:
Ok sounds reasonable to me. I Agree that allowing input range to be given in hz would entail a lot of breaking changes

Reply:
New version pulled. The parameter 'Hz' is now only for the plot, and input/outputs are always in 
rad/sec.

Changing this behavior entaild indeed several changes and should be discussed separately.